### ISSUE TITLE
Running Brave on custom solution for preserving metadata between threads

### ISSUE THREAD

**Issue by pbetkier** [openzipkin/brave#90]
Hello, 

we're planning to adopt Brave at our company and there is one issue we try to tackle to make it possible. 

We already have our own infrastructure for keeping request metadata in thread locals and transferring the information when request processing switches threads. What we'd like to do is to keep span state in our thread locals and make it available to Brave classes that operate on it. From what I see, creating our custom implementation of `ServerAndClientSpanState` that would delegate to our infrastructure and passing it to `ServerTracer` and `ClientTracer` could solve the problem. Unfortunately, `ServerAndClientSpanState` interface is not public.

Could you recommend how to approach running Brave on existing infrastructure? What do you think of making `ServerAndClientSpanState` interface public?

**Comment by kristofa** [comment]
Hello,

Yes I believe you should be able to have your own implementation of `ServerAndClientSpanState`. 

We discussed a while ago that having a common `state` implementation that is shared across implementations (zipkin, brave,...) would make it possible and easier to mix implementations (eg server using brave calling Finagle client for outgoing request).

I'll make the interface public right away.

**Comment by kristofa** [comment]
It is public on current master. Give it a try and let us know if you run into other issues.
As I'm approaching the end of the brave 3.0.0 work I had planned to make another alpha or beta release soon. Hopefully end of this week. It would be good if you can give this release a try.

**Comment by pbetkier** [comment]
Excellent, thanks for such quick help!

**Comment by pbetkier** [comment]
In order to pass my custom implementation to `ServerTracer` and `ClientTracer` I'll need `build()` methods of their builders to be public as well :)

**Comment by kristofa** [comment]
@pbetkier I would propose the following. What if you for now develop against a local copy / branch of brave and when things work out you can submit a pr with required changes?

**Comment by pbetkier** [comment]
I'll do so, that should work better.

**Comment by adriancole** [comment]
ps you might be interested in this topic https://docs.google.com/document/d/10S3wGxV-s2wFeXFE5ZrPAxYUW_HVJ-kUvxW6xG6_A74/edit We discussed it at a working group a couple months ago. There's a follow-up in October if you are interested. https://docs.google.com/document/d/1nFFoIFKsaL5u_XV5k3VpXzFvM2fr2ysYgmKSlFQNJi0/edit

**Comment by kristofa** [comment]
@pbetkier I did quite some work on brave 3.0.0 over the weekend and I think all features I wanted to implement are done.  I plan to release a RC candidate later today.  I also made sure you can configure your own `ServerAndClientSpanState` and provided 

```java
public static final ServerSpan EMPTY = ServerSpan.create(null);
```

in `ServerSpan` as you suggested.  Have a look at `brave-core` readme because the way to initialize the brave api has changed and is a lot nicer now in my opinion.  Let me know if this works out for you or not.

**Comment by kristofa** [comment]
@pbetkier I just released [brave 3.0.0-rc-1](https://github.com/openzipkin/brave/releases/tag/brave-3.0.0-rc-1).  If you can, try it out and give feedback! Thanks!

**Comment by pbetkier** [comment]
@kristofa I've just switched to brave-3.0.0-rc-1 and it does have everything I need for integrating brave into my existing infrastructure. The new API for configuring brave (with a builder) was a good move, simplified setting it up a lot.

I'm closing this issue then, thanks!
